test-data/unit/check-typeis.test

```diff
@@ -0,0 +1,756 @@
+
+[case testAssignToTypeIsedVariable1]
+from typing_extensions import TypeIs
+
+class A: pass
+class B(A): pass
+
+def guard(a: A) -> TypeIs[B]:
+    pass
+
+a = A()
+if not guard(a):
+    a = A()
+[builtins fixtures/tuple.pyi]
+
+[case testAssignToTypeIsedVariable2]
+from typing_extensions import TypeIs
+
+class A: pass
+class B: pass
+
+def guard(a: object) -> TypeIs[B]:
+    pass
+
+a = A()
+if not guard(a):
+    a = A()
+[builtins fixtures/tuple.pyi]
+
+[case testAssignToTypeIsedVariable3]
+from typing_extensions import TypeIs
+
+class A: pass
+class B: pass
+
+def guard(a: object) -> TypeIs[B]:
+    pass
+
+a = A()
+if guard(a):
+    reveal_type(a)  # N: Revealed type is "__main__.<subclass of "A" and "B">"
+    a = B()  # E: Incompatible types in assignment (expression has type "B", variable has type "A")
+    reveal_type(a)  # N: Revealed type is "__main__.<subclass of "A" and "B">"
+    a = A()
+    reveal_type(a)  # N: Revealed type is "__main__.A"
+reveal_type(a)  # N: Revealed type is "__main__.A"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsNestedRestrictionAny]
+from typing_extensions import TypeIs
+from typing import Any
+
+class A: ...
+def f(x: object) -> TypeIs[A]: ...
+def g(x: object) -> None: ...
+
+def test(x: Any) -> None:
+    if not(f(x) or x):
+        return
+    g(reveal_type(x))  # N: Revealed type is "Union[__main__.A, Any]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsNestedRestrictionUnionOther]
+from typing_extensions import TypeIs
+from typing import Any
+
+class A: ...
+class B: ...
+def f(x: object) -> TypeIs[A]: ...
+def f2(x: object) -> TypeIs[B]: ...
+def g(x: object) -> None: ...
+
+def test(x: object) -> None:
+    if not(f(x) or f2(x)):
+        return
+    g(reveal_type(x))  # N: Revealed type is "Union[__main__.A, __main__.B]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsComprehensionSubtype]
+from typing import List
+from typing_extensions import TypeIs
+
+class Base: ...
+class Foo(Base): ...
+class Bar(Base): ...
+
+def is_foo(item: object) -> TypeIs[Foo]:
+    return isinstance(item, Foo)
+
+def is_bar(item: object) -> TypeIs[Bar]:
+    return isinstance(item, Bar)
+
+def foobar(items: List[object]):
+    a: List[Base] = [x for x in items if is_foo(x) or is_bar(x)]
+    b: List[Base] = [x for x in items if is_foo(x)]
+    c: List[Bar] = [x for x in items if is_foo(x)]  # E: List comprehension has incompatible type List[Foo]; expected List[Bar]
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsNestedRestrictionUnionIsInstance]
+from typing_extensions import TypeIs
+from typing import Any, List
+
+class A: ...
+def f(x: List[Any]) -> TypeIs[List[str]]: ...
+def g(x: object) -> None: ...
+
+def test(x: List[Any]) -> None:
+    if not(f(x) or isinstance(x, A)):
+        return
+    g(reveal_type(x))  # N: Revealed type is "Union[builtins.list[builtins.str], __main__.<subclass of "list" and "A">]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypeIsMultipleCondition-xfail]
```

Comment:
I copied this from the TypeGuard test. Seems like the xfail is just because we generate a name with an extra `1` at the end, which looks like a typo:

```
Expected:
  main:16: note: Revealed type is "__main__.<subclass of "Foo" and "Bar">"
  main:21: note: Revealed type is "__main__.<subclass of "Foo" and "Bar">" (diff)
Actual:
  main:16: note: Revealed type is "__main__.<subclass of "Foo" and "Bar">"
  main:21: note: Revealed type is "__main__.<subclass of "Foo" and "Bar">1" (diff)
```

I will remove the xfail and make it pass.